DragControls: Add `rotate` mode.

lewibs opened the pull request:
**Description**

I often use three extensions that do not work well with the TranslateControls. I as a result have been using DragControls but when I need to rotate I have been using my own variation RotateControls. I am pushing this so that perhaps others don't need to rewrite the same code.

It intentionally has extremely similar API to DragControls as well so most of this was copy pasted from that code.

The way that it works, is that when the user clicks on an object, a up vector and horizon are created. When the user moves their mouse, the object is rotated around those vectors.

lewibs commented:
Yes sorry about that. Just reverted it. It should contain the build files. 

I'm talking about TransformControls.
https://threejs.org/docs/#examples/en/controls/TransformControls

Mugen87 commented:
> Yes sorry about that. Just reverted it. It should contain the build files.

Please make to not update build files at all. There should be a diff.

Mugen87 commented:
The feature is actually quite nice! I wonder if it makes sense to add a separate controls class though. 

How about we add a new property to `DragControls` called `mode`? The default would be `translate` (similar to `TransformControls`) but you can set it to `rotate` to enable the new interaction presented in this PR.

[misc_controls_drag](https://threejs.org/examples/misc_controls_drag) can be modified such that users can choose the transformation mode for better testing.

lewibs commented:
I'm curious, what is it about it that you like? I also made a flag for the TransformControls that allows you to use it without the gizmos. It works the same way as the drag and rotate controls combined. However it doesn't change the rotation axis unfortunately it still rotates with the plane instead of about the horizon. I can make two prs if you are interested in seeing the other.

lewibs commented:
I just finished moving the rotation controls into the DragControls and updated the docs and the demo. 

I reverted back to the original build so I'm not sure why its still showing a diff. What would you recommend doing to remove these changes from the history?

Mugen87 commented:
I suggest you revert all your commits in your branch (via `git reset` since this will keep the changes), undo changes to the build files, make a single new commit and then force-push. This will correct the history.

Mugen87 commented on examples/jsm/controls/DragControls.js:
```diff
@@ -12,12 +12,21 @@ const _raycaster = new Raycaster();
+
 class DragControls extends EventDispatcher {
 
+	mode = "translate";
```
Please move this code into the constructor.

lewibs replied:
done

lewibs committed:
added rotation to DragControls

lewibs commented:
Hey just finished that up. It should be good now. Let me know if there are any issues. 

Also let me know if modifying TransformControls to have a flag called gizmoless that would make it act like DragControls would be useful.

Mugen87 commented on examples/jsm/controls/DragControls.js:
```diff
@@ -161,9 +184,19 @@ class DragControls extends EventDispatcher {
+					
+						_up.set(0,1,0).applyQuaternion(_camera.quaternion).normalize();
+						_horizen.set(1,0,0).applyQuaternion(_camera.quaternion).normalize();
+						_lookAt.set(0,0,-1).applyQuaternion(_camera.quaternion).normalize();
```
`_lookAt` is never used. I think this variable can bed deleted.

Mugen87 committed:
Update misc_controls_drag.html

Clean up.

Mugen87 commented on examples/jsm/controls/DragControls.js:
```diff
@@ -75,17 +84,31 @@ class DragControls extends EventDispatcher {
-
-					_selected.position.copy( _intersection.sub( _offset ).applyMatrix4( _inverseMatrix ) );
-
+					_diff.multiplyScalar(2); //Just a random scalling factor since it rotated too slow
```
Instead of using a fixed number, let's add a `rotateSpeed` property like in `OrbitControls` with `1` as the default value.

Mugen87 commented on examples/jsm/controls/DragControls.js:
```diff
@@ -161,9 +184,19 @@ class DragControls extends EventDispatcher {
+					
+					} else if (scope.mode === "rotate") {
+					
+						_up.set(0,1,0).applyQuaternion(_camera.quaternion).normalize();
```
Let's add a comment that the controls only support `Y+` up.

lewibs committed:
updated code per code review see #27689 for details

lewibs committed:
Merge branch 'dev' of https://github.com/lewibs/three.js into dev

Mugen87 committed:
Update DragControls.js

Clean up.

Mugen87 committed:
Update DragControls.html

Clean up.

Mugen87 committed:
Update DragControls.html

Fix type.

Mugen87 committed:
Update DragControls.js

Mugen87 committed:
Update DragControls.js

Mugen87 commented on examples/jsm/controls/DragControls.js:
```diff
@@ -75,14 +85,27 @@ class DragControls extends EventDispatcher {
 			updatePointer( event );
+			_end.copy( _pointer );
+			_diff.subVectors( _end, _start );
+			_start.copy( _end );
```
I think this bit can be changed so we avoid unnecessary computations. The `_diff` computation should only be done inside the `scope.mode === 'rotate'` block. At the end of `onPointerDown()` and `onPointerMove()` if would do this:
```js
_previousPointer.copy( _pointer );
```
And `_diff` is then computed like so:
```js
_diff.subVectors( _pointer, _previousPointer );
```

Mugen87 committed:
Update DragControls.js

Mugen87 commented:
Let's give this a try!

lewibs commented:
Awesome! Thanks so much
